File: 2022/day15.pl
use v5.36;
use Data::Dumper::Concise;

open my $in, '<', 'input/day15.txt' or die "bad open: $!";
my $TARGET = 2000000;
my $MAX    = 4000000;

my %beacons;
my %sensors;
my $num = qr{-?[0-9]+};

# read the grid
while (my $line = <$in>) {
  my ($sx, $sy, $bx, $by) = $line =~ /x=($num), y=($num):.*x=($num), y=($num)/;

  # manhattan distance, then increase the coverage over its whole area
  my $dist = abs($sx - $bx) + abs($sy - $by);
  $sensors{"$sx,$sy"} = $dist;
  $beacons{"$bx,$by"} = 1;
}

close $in;

# PART 1
{
  my %pos;

  for my $k (keys %sensors) {
    my $dist = $sensors{$k};
    my ($sx, $sy) = split /,/, $k;

    my $dist_at_target = $dist - abs($sy - $TARGET);
    next if $dist_at_target <= 0;

    my $startx = $sx - ($dist_at_target);
    my $endx   = $sx + ($dist_at_target);
    for (my $x = $startx; $x <= $endx; $x++) {
      $pos{$x} = 1 unless $beacons{"$x,$TARGET"};
    }
  }

  say "part 1: " . scalar keys %pos;
}

# PART 2
# keys are y values, values are [[ax1,ax2], [bx1,bx2]]
my %blocked;

for my $k (keys %sensors) {
  # say "processing $k to discver blocked things";
  my $dist = $sensors{$k};
  my ($sx, $sy) = split /,/, $k;

  for (my $y = $sy - $dist; $y <= $sy + $dist; $y++) {
    my $xrange = $dist - abs($sy - $y);
    push $blocked{$y}->@*, [$sx - $xrange, $sx + $xrange];
  }
}

OUTER: for my $y (keys %blocked) {
  next if $y < 0 || $y > $MAX;

  my $pairs = delete $blocked{$y};
  my @sorted = sort {; $a->[0] <=> $b->[0] || $a->[1] <=> $b->[1] } @$pairs;

  my $right = 0;    # rightmost filled point on this line

  for my $pair (@sorted) {
    $right = $pair->[1] if $pair->[0] <= $right + 1 <= $pair->[1];
    next OUTER if $right > $MAX;
  }

  # here, the point just right of our rightmost filled spot is empty!
  $right++;
  say "part 2: " . ($right * 4_000_000 + $y);
  last;
}

File: 2022/src/bin/day15.rs
use regex::Regex;
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{prelude::*, BufReader};

type Coord = (isize, isize);

fn main() {
  let reader = BufReader::new(File::open("input/day15.txt").expect("bad open"));

  let mut beacons = HashSet::new();
  let mut sensors = HashMap::new();

  let re =
    Regex::new(r"x=([-0-9]+), y=([-0-9]+):.*x=([-0-9]+), y=([-0-9]+)").unwrap();

  for line in reader.lines().map(|l| l.unwrap().trim().to_string()) {
    let matches = re.captures(&line).unwrap();

    let sensor: Coord = (matches[1].parse().unwrap(), matches[2].parse().unwrap());
    let beacon: Coord = (matches[3].parse().unwrap(), matches[4].parse().unwrap());

    let dist = (sensor.0 - beacon.0).abs() + (sensor.1 - beacon.1).abs();
    beacons.insert(beacon);
    sensors.insert(sensor, dist);
  }

  part1(&sensors, &beacons);
  part2(&sensors);
}

fn part1(sensors: &HashMap<Coord, isize>, beacons: &HashSet<Coord>) {
  let target = 2_000_000;

  let mut pos = HashSet::new();

  for ((x, y), dist) in sensors.iter() {
    let dt = dist - (y - target).abs();

    if dt <= 0 {
      continue;
    }

    let start = x - dt;
    let end = x + dt;
    for this_x in start..=end {
      let tup = (this_x, target);
      if !beacons.contains(&tup) {
        pos.insert(tup);
      }
    }
  }

  println!("part 1: {}", pos.len());
}

fn part2(sensors: &HashMap<Coord, isize>) {
  let max = 4_000_000;

  // keys are y values, values are [[ax1,ax2], [bx1,bx2]]
  let mut blocked = HashMap::new();

  for ((x, y), dist) in sensors.iter() {
    for this_y in (y - dist)..=(y + dist) {
      let xrange = dist - (y - this_y).abs();
      if !blocked.contains_key(&this_y) {
        blocked.insert(this_y, vec![]);
      }

      let vals = blocked.get_mut(&this_y).expect("no key for {y}");
      vals.push((x - xrange, x + xrange));
    }
  }

  'outer: for (y, pairs) in blocked.iter_mut() {
    if y < &&0 || y > &&max {
      continue;
    }

    pairs.sort_unstable_by_key(|pair| pair.0);

    let mut right = 0; // rightmost point on this line

    for (start, end) in pairs.iter() {
      if start <= &(right + 1) && &(right + 1) <= end {
        right = *end;

        if right > max {
          continue 'outer;
        }
      }
    }

    // all done!
    println!("part 2: {}", ((right + 1) * 4_000_000) + y);
    return;
  }
}
